lsp/lsp.go

```diff
@@ -0,0 +1,153 @@
+package lsp
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	_ "github.com/tliron/commonlog/simple"
+	"github.com/tliron/glsp"
+	protocol "github.com/tliron/glsp/protocol_3_16"
+	glspServer "github.com/tliron/glsp/server"
+	"github.com/tliron/kutil/version"
+
+	"github.com/TBD54566975/ftl/go-runtime/compile"
+	ftlErrors "github.com/TBD54566975/ftl/internal/errors"
+)
+
+const lsName = "ftl-language-server"
+
+// Server is a language server.
+type Server struct {
+	Server      *glspServer.Server
+	GlspLogger  *GLSPLogger
+	glspContext *glsp.Context
+	handler     protocol.Handler
+}
+
+// NewServer creates a new language server.
+func NewServer(ctx context.Context) *Server {
+	handler := protocol.Handler{
+		Initialized: initialized,
+		Shutdown:    shutdown,
+		SetTrace:    setTrace,
+		LogTrace:    logTrace,
+	}
+	s := glspServer.NewServer(&handler, lsName, false)
+	server := &Server{
+		Server:     s,
+		GlspLogger: NewGLSPLogger(s.Log),
+	}
+	handler.Initialize = server.initialize()
+	// handler.TextDocumentDidOpen = server.textDocumentDidOpen()
+	handler.TextDocumentDidChange = server.textDocumentDidChange()
+	return server
+}
+
+func (s *Server) Run() error {
+	return errors.Wrap(s.Server.RunStdio(), "lsp")
+}
+
+type errSet map[string]compile.Error
+
+func (s *Server) Post(err error) {
+	errByFilename := make(map[string]errSet)
+
+	// Deduplicate and associate by filename.
+	for _, subErr := range ftlErrors.UnwrapAll(err) {
+		var ce compile.Error
+		if errors.As(subErr, &ce) {
+			cp := ce.Pos
+			if errByFilename[cp.Filename] == nil {
+				errByFilename[cp.Filename] = errSet{}
+			}
+			errByFilename[cp.Filename][strings.TrimSpace(ce.Error())] = ce
+		}
+	}
+
+	go func() {
+		for filename, errs := range errByFilename {
+			var diagnostics []protocol.Diagnostic
+			for _, e := range errs {
+				pp := e.Pos
+				sourceName := "ftl"
+				severity := protocol.DiagnosticSeverityError
+				diagnostics = append(diagnostics, protocol.Diagnostic{
+					Range: protocol.Range{
+						Start: protocol.Position{Line: protocol.UInteger(pp.Line), Character: protocol.UInteger(pp.Column + pp.Offset)},
+						End:   protocol.Position{Line: protocol.UInteger(pp.Line), Character: protocol.UInteger(pp.Column + pp.Offset + 10)}, //todo: fix
```

Comment:
We should introduce a `EndCol` field into the positional information at some point.